### PR TITLE
Realit.io integration

### PR DESCRIPTION
**Pull request by kivutar**
https://github.com/WeTrustPlatform/move-ethereum-forward/issues/761

### CODE REVIEW

File: contracts/CharityChallenge.sol [hunk 1]
```diff
@@ -61,6 +66,7 @@ contract CharityChallenge {
         address payable[] memory _npoAddresses,
         uint8[] memory _ratios,
         address _marketAddress,
```
Comment (tuyennhv) [inline contracts/CharityChallenge.sol]: should we remove `_marketAddress`?

Comment (kivutar) [reply contracts/CharityChallenge.sol]: We need to limit ABI changes. Also, this _marketAddress will be now pointing to the Realit.io contract.

**Comment by tuyennhv** [comment]
If possible, I'd like to take this chance to add more utilitity methods so that we don't have to make multiple web3 calls. One of them is to determine challenge step like this:
```
export const challengeStep = (challenge, finalized, accomplished, today) => {
  const endTimeInMillis = Date.parse(challenge.endTime);
  if (!challenge.contractAddress) return 'draft';

  let step = 'unresolved';
  const now = today || Date.now();
  if (endTimeInMillis > now) {
    step = 'pledging';
  } else if (finalized && accomplished) {
    step = 'success';
  } else if (finalized && !accomplished) {
    step = 'failed';
  } else if (!finalized) {
    const SAFETY_HATCH_IN_WEEK = 26;
    if (
      now <
      endTimeInMillis + SAFETY_HATCH_IN_WEEK * 7 * 24 * 60 * 60 * 1000
    ) {
      step = 'unresolved';
    } else {
      step = 'failed';
    }
  }
  return step;
};
```

File: contracts/CharityChallenge.sol [hunk 2]
```diff
@@ -73,10 +77,12 @@ contract CharityChallenge {
         address _arbitrator,
         uint256 _timeout,
         uint256 _challengeEndTime,
+        uint8 _makerFee,
         bool _unlockOnNo
     ) public
     {
         require(_npoAddresses.length == _ratios.length);
+        require(_makerFee < 100);
```
Comment (sihoang) [inline contracts/CharityChallenge.sol]: I think we should use a finer grain to support 2 decimal points like the standard credit card fee 2.99%
So the value of 1% is 100
_makerFee < 10000

https://www.reddit.com/r/ethereum/comments/3s6rfh/best_practices_for_handling_percentages_in/

File: contracts/CharityChallenge.sol [hunk 3]
```diff
@@ -133,10 +140,17 @@ contract CharityChallenge {
                 uint256 totalContractBalance = address(this).balance;
                 uint length = npoAddresses.length;
                 uint256 donatedAmount = 0;
+                if (makerFee > 0) {
+                    uint256 amount = totalContractBalance * makerFee / 100;
```
Comment (sihoang) [inline contracts/CharityChallenge.sol]: if we use the above `part-per` then 100 becomes 10000

File: contracts/CharityChallenge.sol [hunk 4]
```diff
@@ -141,22 +144,22 @@ contract CharityChallenge {
...
-                    uint256 amount = amountForNPO * ratio / sumRatio;
-                    donatedAmount += amount;
+                    uint256 amount = amountForNPO.mul(ratio).div(sumRatio);
+                    donatedAmount = donatedAmount.add(amount);
                     npo.transfer(amount);
                     emit Donated(npo, amount);
                 }
                 // Don't want to keep any amount in the contract
-                uint256 remainingAmount = totalContractBalance - donatedAmount;
+                uint256 remainingAmount = totalContractBalance.sub(donatedAmount);
```
Comment (knd) [inline contracts/CharityChallenge.sol]: * is there a chance that `donatedAmount` is greater than `totalContractBalance` (due to precision through calculations above)? If not, it's ok to leave as is. If yes, would the current code statements or the next statements throw error? 

* can we use `address(this).balance` to know the remaining balance here without doing calculation, as the transfer statements have been already executed above? 

* if we can use `address(this).balance` after all the transfer statements, i guess we don't really need to have variable `donatedAmount` for bookkeeping.

Comment (kivutar) [reply contracts/CharityChallenge.sol]: I've tried to use `address(this).balance` here and it breaks the tests. I suppose that the balance of the contract is updated only at the end of the function call.

Comment (kivutar) [reply contracts/CharityChallenge.sol]: Yep, it simplified the code greatly, thanks!

**Pull request merged**